### ISSUE TITLE
Program Throws an Exception with UDQ

### ISSUE THREAD

**Issue by OPMUSER** [OPM/opm-simulators#2638]
The following code in the input deck cause an exception.
```
UDQ 
-- 
-- ASSIGN CONSTANTS
--             
ASSIGN FUTEMP     225.0                       / Reservoir Temperature   

ASSIGN FUNGLYLD   1.1963929             / Condensate  Yield (63.5/56.766)/(1.0 - 0.065)
/
```
Program ends with:
```
Failed to create valid EclipseState object.
Exception caught:stod
```
Any ideas?

**Comment by joakim-hove** [comment]
It is the 
```
ASSIGN FUNGLYLD   1.1963929             / Condensate  Yield (63.5/56.766)/(1.0 - 0.065)
```
which is the culprit; the parser will consider everything following the *last* `/` as comment; that will be wrong here. This works:
```
ASSIGN FUNGLYLD   1.1963929             / -- Condensate  Yield (63.5/56.766)/(1.0 - 0.065)
```

**Comment by OPMUSER** [comment]
@joakim-hove okay I can work around that for this release. But OPM Flow should ignore everything after the first `/` as per the commercial simulator.

Good support though :-)

**Comment by joakim-hove** [comment]
> But OPM Flow should ignore everything after the first / as per the commercial simulator.

yes opm/flow generally does it correct; i.e. this
```
TABDIMS
  1 2 3 / And here comes a comment with / in it
```
works fine - however for the *two* keywords `UDQ` and `ACTIONX` where the first `/` might be part of a mathematical expression we do it differently - e.g here:
```
UDQ
     DEFINE WUWCT WWPR '*' / (WWPR '*' + WOPR '*') /
/ 
```
you can cut at the first `/`.

**Comment by OPMUSER** [comment]
The logic is not robust as the commercial simulator has no problem with my ~180 line script::
```

UDQ 
-- 
-- ASSIGN CONSTANTS
--          
ASSIGN FUTEMP     225.0                                                        / Reservoir Temperature                               
ASSIGN FUNGLYLD   1.1963929                                                    / Condensate  Yield (63.5/56.766)/(1.0 - 0.065)
-- ASSIGN FUNGLSHK   0.0650                                                    / Condensate  Shrinkage Factor
-- ASSIGN FUNGLSHK   0.0069                                                    / Condensate  Shrinkage Factor (Adjusted)
ASSIGN FUNGLSHK   0.000000                                                     / Condensate  Shrinkage Factor set to Zero
ASSIGN FULPGYLD   0.065775                                                     / LPG Sep Gas Yield (stb/Mscf) 
ASSIGN FULPGSHK   0.080410                                                     / LPG         Shrinkage Factor (0.0860)*(1.0 - 0.065)
ASSIGN FUFACSHK   0.000935                                                     / Facilities  Shrinkage Factor (0.0010)*(1.0 - 0.065)
ASSIGN FUFULSHK   0.052924528301887                                            / Fuel        Utilization   (106.0/6.0)*(1.0 - 0.065)
ASSIGN FUDELTA    1E-10                                                        / Value to avoid diving by zero errors                                                     
-- 
-- ASSIGN FUCONFAC   1.000000                                                  / Conversion Factor for Eclipse
ASSIGN FUCONFAC   5.614583E-3                                                  / Conversion Factor for tNavigator

ASSIGN WUWEFAC    0.95                                                         / Well Efficiency Factor
ASSIGN WUDELTA    1E-10                                                        / Value to avoid diving by zero errors                                                     
--
-- SET INITIAL VALUES VALUES
--
ASSIGN FU_FGST    0.0                                                          / Gas Sales  Total

ASSIGN FU_FLPGT   0.0                                                          / LPG        Total
ASSIGN WU_WLPGT   0.0                                                          / LPG        Total
                    
ASSIGN FU_FNGLT   0.0                                                          / Condensate Total
ASSIGN WU_WNGLT   0.0                                                          / Condensate Total

ASSIGN FU_LOGIC   0                                                            / Gas Sales Logical Flag
                                                                                                      
DEFINE DRY AND WET GAS VOLUMES   
      
DEFINE WU_WDGPR (FUCONFAC * WTPRIGS)                                           / Calculate Dry Gas Rate
UPDATE WU_WDGPR ON                                                             / Need to convert from STB to MSCF for tNavigator
UNITS  WU_WDGPR MSCFD                                                          / 
             
DEFINE WU_WDGPT (FUCONFAC * WTPTIGS)                                           / Calculate Dry Gas Total    Well (tNavigator)
UPDATE WU_WDGPT ON                                                             / Need to convert from STB to MSCF  for tNavigator
UNITS  WU_WDGPT MSCF                                                           /

DEFINE WU_WWGPR WGPR - WU_WDGPR                                                / Calculate Wet Gas Rate     Well
UPDATE WU_WWGPR ON                                                             /
UNITS  WU_WWGPR MSCFD                                                          /

DEFINE WU_WWGPT WGPT - WU_WDGPT                                                / Calculate Wet Gas Total    Well
UPDATE WU_WWGPT ON                                                             /
UNITS  WU_WWGPT MSCF                                                           /

DEFINE FU_FDGPR (FUCONFAC * FTPRIGS)                                           / Calculate Dry Gas Rate     Field
UPDATE FU_FDGPR ON                                                             / 
UNITS  FU_FDGPR MSCFD                                                          /
                                                          
DEFINE FU_FDGPT (FUCONFAC * FTPTIGS)                                           / Calculate Dry Gas Total    Field
UPDATE FU_FDGPT ON                                                             /
UNITS  FU_FDGPT MSCF                                                           /

DEFINE FU_FWGPR FGPR - FU_FDGPR                                                / Calculate Wet Gas Rate     Field
UPDATE FU_FWGPR ON                                                             / 
UNITS  FU_FWGPR MSCFD                                                          /
                                              
DEFINE FU_FWGPT FGPT - FU_FDGPT                                                / Calculate Dry Gas Total    Field
UPDATE FU_FWGPT ON                                                             /
UNITS  FU_FWGPT MSCF                                                           /
--
--
-- DEFINE CONDENSATE VARIABLES
--
DEFINE WU_WNGLR WGPR *(WOGR * FUNGLYLD)                                        / Calculate Condensate Rate  Well
UPDATE WU_WNGLR ON                                                             /
UNITS  WU_WNGLR STBD                                                           /

DEFINE WU_WNGLT (WOPT * FUNGLYLD)                                              / Calculate Condensate Total Well
UPDATE WU_WNGLT ON                                                             /
UNITS  WU_WNGLT STB                                                            /

DEFINE WU_WNGLY WU_WNGLR / (WGPR + WUDELTA)                                    / Calculate Condensate Yield Well
UPDATE WU_WNGLY ON                                                             /
UNITS  WU_WNGLY STBMSCF                                                        /
            
DEFINE FU_FNGLR FGPR *(FOGR * FUNGLYLD)                                        / Calculate Condensate Rate  Field
UPDATE FU_FNGLR ON                                                             /
UNITS  FU_FNGLR STBD                                                           /
                                              
DEFINE FU_FNGLT (FOPT * FUNGLYLD)                                              / Calculate Condensate Total Field)
UPDATE FU_FNGLT ON                                                             /
UNITS  FU_FNGLT STB                                                            /

DEFINE FU_FNGLY FU_FNGLR / (FGPR + FUDELTA)                                    / Calculate Condensate Yield Field
UPDATE FU_FNGLY ON                                                             /
UNITS  FU_FNGLY STBMSCF                                                        /
--
-- DEFINE LPG VARIABLES                        
--
DEFINE WU_WLPGR WU_WWGPR * FULPGYLD                                            / Calculate LPG Rate  Well
UPDATE WU_WLPGR ON                                                             /
UNITS  WU_WLPGR STBD                                                           /

DEFINE WU_WLPGT WU_WWGPT * FULPGYLD                                            / Calculate LPG Total Well
UPDATE WU_WLPGT ON                                                             / Convert to Mscf
UNITS  WU_WLPGT STB                                                            /

DEFINE WU_WLPGY WU_WLPGR / (WGPR + WUDELTA)                                    / Calculate LPG Yield Well
UPDATE WU_WLPGY ON                                                             /
UNITS  WU_WLPGY STBMSCF                                                        /

DEFINE FU_FLPGR FU_FWGPR * FULPGYLD                                            / Calculate LPG Rate  Field
UPDATE FU_FLPGR ON                                                             /
UNITS  FU_FLPGR STBD                                                           /

DEFINE FU_FLPGT FU_FWGPT * FULPGYLD                                            / Calculate LPG Total Field
UPDATE FU_FLPGT ON                                                             /
UNITS  FU_FLPGT STB                                                            /

DEFINE FU_FLPGY FU_FLPGR / (FGPR + FUDELTA)                                    / Calculate LPG Yield Field
UPDATE FU_FLPGY ON                                                             /
UNITS  FU_FLPGY STBMSCF                                                        /
--                                                                                                       
-- DEFINE GAS RE-INECTION RATE VARIABLES AND SALES GAS
--                                               
DEFINE FUNGLFAC 1.00                                                           / Define Condensate Weight Factor 
DEFINE FULPGFAC 1.00                                                           / Define LPG        Weight Factor
DEFINE FUNGLLOS FU_FWGPR * FUNGLFAC * FUNGLSHK                                 / Define Condensate Gas Shrinkage
DEFINE FULPGLOS FU_FWGPR * FULPGFAC * FULPGSHK                                 / Define LPG        Gas Shrinkage
DEFINE FUFACLOS FU_FWGPR * FUFACSHK                                            / Define Facilities Gas Shrinkage
DEFINE FU_GASNR FU_FWGPR + FU_FDGPR - FUNGLLOS - FULPGLOS - FUFACLOS           / Define Post Processing Gas Rate
DEFINE FU_FUELR FU_GASNR * FUFULSHK                                            / Define Fuel       Gas Shrinkage

DEFINE FU_GIR   (FU_GASNR - FU_FUELR) * (1 - FU_LOGIC)                         / Define Gas Re-Injection Rate 
UPDATE FU_GIR   ON                                                             /
UNITS  FU_GIR   MSCF                                                           /

DEFINE FU_FGSR  (FU_GASNR - FU_FUELR) * FU_LOGIC                               / Define Gas Sales Rate 
UPDATE FU_FGSR  ON                                                             /
UNITS  FU_FGSR  MSCF                                                           /
--                                    
-- DEFINE GAS RE-INECTION TOTAL VARIABLES
--                                               
DEFINE FUNGLLOT FU_FWGPT * FUNGLFAC * FUNGLSHK                                 / Define Condensate Gas Shrinkage
DEFINE FULPGLOT FU_FWGPT * FULPGFAC * FULPGSHK                                 / Define LPG        Gas Shrinkage
DEFINE FUFACLOT FU_FWGPT * FUFACSHK                                            / Define Facilities Gas Shrinkage
DEFINE FU_GASNT FU_FWGPT + FU_FDGPT - FUNGLLOT - FULPGLOT - FUFACLOT           / Define Post Processing Gas Rate
DEFINE FU_FUELT FU_GASNT * FUFULSHK                                            / Define Fuel       Gas Shrinkage   

DEFINE FU_FGST   (FU_GASNT - FU_FUELT - FGIT) * FU_LOGIC                       / Define Gas Sales Total 
UPDATE FU_FGST   ON                                                            /
UNITS  FU_FGST   MSCF                                                          /
                                 
/                                                                                End of UDQ Keyword

-- ---------------------------------------------------------------------------------------------------------------------------------
-- CALCULATE FIELD GROUP GAS RE-INJECTION RATE 
-- ---------------------------------------------------------------------------------------------------------------------------------
ACTIONX
        GRUPINJ         10000 30                                               /        
        FGIR          > 0                                                      /
/ 

-- GROUP INJECTION CONTROLS                                                     
--                                                                              
-- GRUP   FLUID CNTL   SURF   RESV   REINJ  VOID   GRUP  GUIDE  GUIDE GRUP  GRUP
-- NAME   TYPE  MODE   RATE   RATE   FRAC   FRAC   CNTL  RATE   DEF   REINJ RESV               
GCONINJE                                                                        
'FIELD'  'GAS' 'RATE'  FU_GIR 1*     1*     1*     'YES' 1*     1*    1*    1* /
/                                                                               

ENDACTIO

```
I can work around this as you suggested so not a problem, but perhaps we can improve on this for the 2020-10 release ;-)

**Comment by OPMUSER** [comment]
Still does not work commented out all parameters except the first like this"
```
UDQ 
-- 
-- ASSIGN CONSTANTS
--          
ASSIGN FUTEMP     225.0                                                        / -- Reservoir Temperature                               
-- ASSIGN FUNGLYLD   1.1963929                                                    / -- Condensate  Yield (63.5/56.766)/(1.0 - 0.065)
-- ASSIGN FUNGLSHK   0.0650                                                    / -- Condensate  Shrinkage Factor
-- ASSIGN FUNGLSHK   0.0069                                                    / -- Condensate  Shrinkage Factor (Adjusted)
--                ASSIGN FUNGLSHK   0.000000                                                     / -- Condensate  Shrinkage Factor set to Zero
--                ASSIGN FULPGYLD   0.065775                                                     / -- LPG Sep Gas Yield (stb/Mscf) 
--                ASSIGN FULPGSHK   0.080410                                                     / -- LPG         Shrinkage Factor (0.0860)*(1.0 - 0.065)
--                ASSIGN FUFACSHK   0.000935                                                     / -- Facilities  Shrinkage Factor (0.0010)*(1.0 - 0.065)
--                ASSIGN FUFULSHK   0.052924528301887                                            / -- Fuel        Utilization   (106.0/6.0)*(1.0 - 0.065)
--                ASSIGN FUDELTA    1E-10                                                        / -- Value to avoid diving by zero errors                                                     
--                -- 
--                ASSIGN FUCONFAC   1.000000                                                     / Conversion Factor for Eclipse
--                -- ASSIGN FUCONFAC   5.614583E-3                                                  / Conversion Factor for tNavigator
--                
--                ASSIGN WUWEFAC    0.95                                                         / Well Efficiency Factor
--                ASSIGN WUDELTA    1E-10                                                        / Value to avoid diving by zero errors                                                     
--                --
--                -- SET INITIAL VALUES VALUES
--                --
--                ASSIGN FU_FGST    0.0                                                          / Gas Sales  Total
--                
--                ASSIGN FU_FLPGT   0.0                                                          / LPG        Total
--                ASSIGN WU_WLPGT   0.0                                                          / LPG        Total
--                                    
--                ASSIGN FU_FNGLT   0.0                                                          / Condensate Total
--                ASSIGN WU_WNGLT   0.0                                                          / Condensate Total
--                
--                ASSIGN FU_LOGIC   0                                                            / Gas Sales Logical Flag
--                --                                                                                                      
--                -- DEFINE DRY AND WET GAS VOLUMES   
--                --      
--                DEFINE WU_WDGPR (FUCONFAC * WTPRIGS)                                           / Calculate Dry Gas Rate
--                -- UPDATE WU_WDGPR ON                                                             / Need to convert from STB to MSCF for tNavigator
--                UNITS  WU_WDGPR MSCFD                                                          / 
--                             
--                DEFINE WU_WDGPT (FUCONFAC * WTPTIGS)                                           / Calculate Dry Gas Total    Well (tNavigator)
--                --- UPDATE WU_WDGPT ON                                                             / Need to convert from STB to MSCF  for tNavigator
--                UNITS  WU_WDGPT MSCF                                                           /
--                
--                DEFINE WU_WWGPR WGPR - WU_WDGPR                                                / Calculate Wet Gas Rate     Well
--                -- UPDATE WU_WWGPR ON                                                             /
--                UNITS  WU_WWGPR MSCFD                                                          /
--                
--                DEFINE WU_WWGPT WGPT - WU_WDGPT                                                / Calculate Wet Gas Total    Well
--                -- UPDATE WU_WWGPT ON                                                             /
--                UNITS  WU_WWGPT MSCF                                                           /
--                
--                DEFINE FU_FDGPR (FUCONFAC * FTPRIGS)                                           / Calculate Dry Gas Rate     Field
--                -- UPDATE FU_FDGPR ON                                                             / 
--                UNITS  FU_FDGPR MSCFD                                                          /
--                                                                          
--                DEFINE FU_FDGPT (FUCONFAC * FTPTIGS)                                           / Calculate Dry Gas Total    Field
--                -- UPDATE FU_FDGPT ON                                                             /
--                UNITS  FU_FDGPT MSCF                                                           /
--                
--                DEFINE FU_FWGPR FGPR - FU_FDGPR                                                / Calculate Wet Gas Rate     Field
--                -- UPDATE FU_FWGPR ON                                                             / 
--                UNITS  FU_FWGPR MSCFD                                                          /
--                                                              
--                DEFINE FU_FWGPT FGPT - FU_FDGPT                                                / Calculate Dry Gas Total    Field
--                -- UPDATE FU_FWGPT ON                                                             /
--                UNITS  FU_FWGPT MSCF                                                           /
--                --
--                --
--                -- DEFINE CONDENSATE VARIABLES
--                --
--                DEFINE WU_WNGLR WGPR *(WOGR * FUNGLYLD)                                        / Calculate Condensate Rate  Well
--                -- UPDATE WU_WNGLR ON                                                             /
--                UNITS  WU_WNGLR STBD                                                           /
--                
--                DEFINE WU_WNGLT (WOPT * FUNGLYLD)                                              / Calculate Condensate Total Well
--                -- UPDATE WU_WNGLT ON                                                             /
--                UNITS  WU_WNGLT STB                                                            /
--                
--                DEFINE WU_WNGLY WU_WNGLR / (WGPR + WUDELTA)                                    / Calculate Condensate Yield Well
--                -- UPDATE WU_WNGLY ON                                                             /
--                UNITS  WU_WNGLY STBMSCF                                                        /
--                            
--                DEFINE FU_FNGLR FGPR *(FOGR * FUNGLYLD)                                        / Calculate Condensate Rate  Field
--                -- UPDATE FU_FNGLR ON                                                             /
--                UNITS  FU_FNGLR STBD                                                           /
--                                                              
--                DEFINE FU_FNGLT (FOPT * FUNGLYLD)                                              / Calculate Condensate Total Field)
--                -- UPDATE FU_FNGLT ON                                                             /
--                UNITS  FU_FNGLT STB                                                            /
--                
--                DEFINE FU_FNGLY FU_FNGLR / (FGPR + FUDELTA)                                    / Calculate Condensate Yield Field
--                -- UPDATE FU_FNGLY ON                                                             /
--                UNITS  FU_FNGLY STBMSCF                                                        /
--                --
--                -- DEFINE LPG VARIABLES                        
--                --
--                DEFINE WU_WLPGR WU_WWGPR * FULPGYLD                                            / Calculate LPG Rate  Well
--                -- UPDATE WU_WLPGR ON                                                             /
--                UNITS  WU_WLPGR STBD                                                           /
--                
--                DEFINE WU_WLPGT WU_WWGPT * FULPGYLD                                            / Calculate LPG Total Well
--                -- UPDATE WU_WLPGT ON                                                             / Convert to Mscf
--                UNITS  WU_WLPGT STB                                                            /
--                
--                DEFINE WU_WLPGY WU_WLPGR / (WGPR + WUDELTA)                                    / Calculate LPG Yield Well
--                -- UPDATE WU_WLPGY ON                                                             /
--                UNITS  WU_WLPGY STBMSCF                                                        /
--                
--                DEFINE FU_FLPGR FU_FWGPR * FULPGYLD                                            / Calculate LPG Rate  Field
--                -- UPDATE FU_FLPGR ON                                                             /
--                UNITS  FU_FLPGR STBD                                                           /
--                
--                DEFINE FU_FLPGT FU_FWGPT * FULPGYLD                                            / Calculate LPG Total Field
--                -- UPDATE FU_FLPGT ON                                                             /
--                UNITS  FU_FLPGT STB                                                            /
--                
--                DEFINE FU_FLPGY FU_FLPGR / (FGPR + FUDELTA)                                    / Calculate LPG Yield Field
--                -- UPDATE FU_FLPGY ON                                                             /
--                UNITS  FU_FLPGY STBMSCF                                                        /
--                --                                                                                                       
--                -- DEFINE GAS RE-INECTION RATE VARIABLES AND SALES GAS
--                --                                               
--                DEFINE FUNGLFAC 1.00                                                           / Define Condensate Weight Factor 
--                DEFINE FULPGFAC 1.00                                                           / Define LPG        Weight Factor
--                DEFINE FUNGLLOS FU_FWGPR * FUNGLFAC * FUNGLSHK                                 / Define Condensate Gas Shrinkage
--                DEFINE FULPGLOS FU_FWGPR * FULPGFAC * FULPGSHK                                 / Define LPG        Gas Shrinkage
--                DEFINE FUFACLOS FU_FWGPR * FUFACSHK                                            / Define Facilities Gas Shrinkage
--                DEFINE FU_GASNR FU_FWGPR + FU_FDGPR - FUNGLLOS - FULPGLOS - FUFACLOS           / Define Post Processing Gas Rate
--                DEFINE FU_FUELR FU_GASNR * FUFULSHK                                            / Define Fuel       Gas Shrinkage
--                
--                DEFINE FU_GIR   (FU_GASNR - FU_FUELR) * (1 - FU_LOGIC)                         / Define Gas Re-Injection Rate 
--                -- UPDATE FU_GIR   ON                                                             /
--                UNITS  FU_GIR   MSCF                                                           /
--                
--                DEFINE FU_FGSR  (FU_GASNR - FU_FUELR) * FU_LOGIC                               / Define Gas Sales Rate 
--                -- UPDATE FU_FGSR  ON                                                             /
--                UNITS  FU_FGSR  MSCF                                                           /
--                --                                    
--                -- DEFINE GAS RE-INECTION TOTAL VARIABLES
--                --                                               
--                DEFINE FUNGLLOT FU_FWGPT * FUNGLFAC * FUNGLSHK                                 / Define Condensate Gas Shrinkage
--                DEFINE FULPGLOT FU_FWGPT * FULPGFAC * FULPGSHK                                 / Define LPG        Gas Shrinkage
--                DEFINE FUFACLOT FU_FWGPT * FUFACSHK                                            / Define Facilities Gas Shrinkage
--                DEFINE FU_GASNT FU_FWGPT + FU_FDGPT - FUNGLLOT - FULPGLOT - FUFACLOT           / Define Post Processing Gas Rate
--                DEFINE FU_FUELT FU_GASNT * FUFULSHK                                            / Define Fuel       Gas Shrinkage   
--                      
--                DEFINE FU_FGST   (FU_GASNT - FU_FUELT - FGIT) * FU_LOGIC                       / Define Gas Sales Total 
--                -- UPDATE FU_FGST   ON                                                            /
--                UNITS  FU_FGST   MSCF                                                          /
                                 
/                                                                                End of UDQ Keyword
```
And got this:
![image](https://user-images.githubusercontent.com/17958927/83326741-011be700-a2a9-11ea-863f-ce06cca9fa3b.png)

**Comment by joakim-hove** [comment]
Ok - can you post the case? I will look into it but not today.

**Comment by OPMUSER** [comment]
Okay your are innocent and I'm afraid I'm guilty. It is crashing because there are a tracer variables in some of the calculations and OPM Flow is not handling tracers; hence the exception. That is the main reason for the issue, there are probably others as well.

Also the `GCONINJE`  keyword is not recognized in the `ACTIONX` statement, which is trapped by the simulator.

The script was being used to calculate the "correct" injection volume taking into account the gas shrinkage due to condensate and LPG extraction. So I cannot get this test case to work - so closing.

**Comment by OPMUSER** [comment]
@joakim-hove this took longer than expected ;-(.

I ended up setting up a series of cases to see what was going on. 

- ACTN01 is the base case run with no ACTIONX or UDQ keywords, just to make sure the case runs. Note there are problems with this case as GP01 goes from 50 MMscf/d to zero, when it should decline based on the BHP constraint - but that is another story.

- ACTN02 has the ACTIONX keyword that switches off gas re-injection and converts GI01 to a producer. This STOPS due to the GCONINJE keyword in the ACTIONX block being unsupported - which is disappointing, but at least this is caught.

- ACTN03 is the same as ACTN02 but with a UDQ section. The UDQ variables are NOT used in this run, but the case ABORTS. This may be because the GCONINJE keyword is not caught in this instance. If you look at the *.PRT file the ACTIONX keyword seems to have been "skipped".

It is challenging to setup test cases for the ACTIONX keyword because of the limited number of keywords supported in the ACTIONX block, so this is the best I can do for the moment.

Cases are attached.

[ACTIONX.zip](https://github.com/OPM/opm-simulators/files/4750224/ACTIONX.zip)

**Comment by alfbr** [comment]
Please consider if these cases are well suited for inclusion in a regression test.

**Comment by OPMUSER** [comment]
I think the first case ACTN01 is suitable for regression testing and in fact the case shows a problem with the current nightly build as shown below:

![image](https://user-images.githubusercontent.com/17958927/84225585-206d0c80-ab12-11ea-89c6-d4a03b86ec67.png)

The red line is the 2020-04 release and that behavior is correct, the nightly shuts in the well instead of producing on the minimum BHP limit.

The ACTN02 and ACTN03 cases are not working at all, so they cannot be used for regression testing at the moment. Ideally, when they are working then we can use then for regression testing. I intend to add one more case to the series (ACTN04) that will use a dry gas tracer to calculate the ratio of dry and wet gas, and then to calculate the LPG yield, LPG shrinkage, and the available gas for re-injection. Currently passive tracers are not working.

The  ACTN04 work flow is based on on an actual gas condensate re-cycling study that was performed using the commercial simulator.

**Comment by joakim-hove** [comment]
OK; I have looked at the third case `ACTN03`. The keywords `UDQ` and `ACTIONX`are treated as special cases in the parser, and for these keywords the parser is generally less powerful - this is biting us here. The limitations regarding these keywords at least include:

1. A record is terminated at the *last* slash of a line - not the first. This is to allow `/` in the mathematical expressions. 

2. A '*' is interpreted as mathematical token which should be part of an expression, and does not invoke the default behavior.

3. This I have not checked in detail - but it seems dos encoded files with `\r\n` line ending are not treated properly.

For the particular case I have:

1. Removed the trailing comments for all the `UDQ` records:
```UDQ
--
-- OPERATOR VARIABLE  EXPRESSION
--        
ASSIGN      FUNGLYLD  1.1963929                             /
ASSIGN      FUNGLSHK  0.000000                              /
ASSIGN      FULPGYLD  0.065775                              /
ASSIGN      FULPGSHK  0.080410                              /
ASSIGN      FUFACSHK  0.000935                              /
ASSIGN      FUFULSHK  0.052924528301887                     /
ASSIGN      FUDELTA   1E-10                                 /
--                     
ASSIGN      FUCONFAC   1.000000                             /
-- ASSIGN FUCONFAC   5.614583E-3                            /
--
DEFINE      FU_FDEPL  FUCONFAC *(50E3)                      /
UNITS       FU_FDEPL  MMSCFD                                /
/
```
2. Converted the file from dos based line endings to unix line endings.

Then the file will parse OK, but it will still stop because the `GCONINJE` keyword is not supported in the `ACTIONX` section.

I had not considered the parsing problems as severe - and was actually quite happy we managed to extend the parser with new behavior for `*` and `/` with special casing too much, but it seems the implementation needs to improve on these matters.
